In [1]:
import pandas as pd
# 데이터 전처리
from mlxtend.preprocessing import TransactionEncoder
# 지지도 계산
from mlxtend.frequent_patterns import apriori
# 연관 규칙
from mlxtend.frequent_patterns import association_rules

In [2]:
ds_asso = pd.read_csv("./실습파일/상품구매.csv", engine = "python")
ds_asso.head()

ID    PRODUCT
0  C-11   BAGUETTE
1  C-11    HERRING
2  C-11    AVOCADO
3  C-11  ARTICHOKE
4  C-11   HEINEKEN

In [8]:
# 고객 ID
ID = list(set(ds_asso["ID"]))
ID.sort()

list_association = []
for i in ID:
    # ID별 구매 데이터 추출
    tmp_list = list(ds_asso[ds_asso["ID"] == i]["PRODUCT"])
    tmp_list.sort()
#     print(tmp_list)
    # ID별 구매 데이터를 리스트에 담기
    list_association.append(tmp_list)
# 리스트 출력
for row in list_association:
    print(row)

['APPLES', 'ARTICHOKE', 'AVOCADO', 'BAGUETTE', 'CORNED BEEF', 'HEINEKEN', 'HERRING']
['APPLES', 'CORNED BEEF', 'HEINEKEN', 'HERRING', 'OLIVES', 'SARDINES', 'STEAK']
['APPLES', 'AVOCADO', 'BAGUETTE', 'ICE CREAM', 'PEPPERS', 'SARDINES', 'STEAK']
['APPLES', 'COKE', 'CORNED BEEF', 'HAM', 'HERRING', 'OLIVES', 'TURKEY']
['ARTICHOKE', 'BOURBON', 'COKE', 'HAM', 'ICE CREAM', 'OLIVES', 'TURKEY']
['ARTICHOKE', 'AVOCADO', 'BAGUETTE', 'COKE', 'HEINEKEN', 'HERRING', 'TURKEY']
['APPLES', 'CHICKEN', 'COKE', 'CORNED BEEF', 'HEINEKEN', 'ICE CREAM', 'SARDINES']
['BAGUETTE', 'BOURBON', 'CRACKERS', 'HEINEKEN', 'OLIVES', 'PEPPERS', 'SODA']
['BOURBON', 'CRACKERS', 'HEINEKEN', 'HERRING', 'OLIVES', 'SODA', 'STEAK']
['APPLES', 'BAGUETTE', 'CORNED BEEF', 'HAM', 'HERRING', 'OLIVES', 'TURKEY']
['ARTICHOKE', 'AVOCADO', 'BAGUETTE', 'BOURBON', 'CORNED BEEF', 'HEINEKEN', 'HERRING']
['ARTICHOKE', 'BOURBON', 'CRACKERS', 'HEINEKEN', 'OLIVES', 'SODA', 'STEAK']
['BOURBON', 'CORNED BEEF', 'CRACKERS', 'HEINEKEN', 'HERRING', 

In [5]:
# Items 중 index 0번이 갖고 있으면 True, 아니면 False
enc = TransactionEncoder()
arr = enc.fit_transform(X = list_association)
# 위의 함수는 데이터를 행렬의 형태로 반환 -> pandas의 DataFrame의 형태로 변환
ds_pre = pd.DataFrame(arr,columns = enc.columns_)
ds_pre

APPLES  ARTICHOKE  AVOCADO  BAGUETTE  BOURBON  CHICKEN   COKE  \
0     True       True     True      True    False    False  False   
1     True      False    False     False    False    False  False   
2     True      False     True      True    False    False  False   
3     True      False    False     False    False    False   True   
4    False       True    False     False     True    False   True   
5    False       True     True      True    False    False   True   
6     True      False    False     False    False     True   True   
7    False      False    False      True     True    False  False   
8    False      False    False     False     True    False  False   
9     True      False    False      True    False    False  False   
10   False       True     True      True     True    False  False   
11   False       True    False     False     True    False  False   
12   False      False    False     False     True    False  False   
13    True      False    False      True    False     True   True   
14   False      False    False     False    False     True   True   
15    True      False    False      True     True    False   True   
16   False      False     True     False     True    False   True   
17   False       True    False     False     True    False  False   
18   False      False    False     False     True    False   True   
19   False      False    False     False     True    False   True   

    CORNED BEEF  CRACKERS    HAM  HEINEKEN  HERRING  ICE CREAM  OLIVES  \
0          True     False  False      True     True      False   False   
1          True     False  False      True     True      False    True   
2         False     False  False     False    False       True   False   
3          True     False   True     False     True      False    True   
4         False     False   True     False    False       True    True   
5         False     False  False      True     True      False   False   
6          True     False  False      True    False       True   False   
7         False      True  False      True    False      False    True   
8         False      True  False      True     True      False    True   
9          True     False   True     False     True      False    True   
10         True     False  False      True     True      False   False   
11        False      True  False      True    False      False    True   
12         True      True  False      True     True      False    True   
13        False     False  False      True    False       True   False   
14         True     False  False      True    False       True   False   
15        False     False  False     False    False       True    True   
16        False     False  False     False    False       True    True   
17        False      True  False      True    False      False    True   
18        False     False  False     False    False       True    True   
19        False     False   True     False    False       True    True   

    PEPPERS  SARDINES   SODA  STEAK  TURKEY  
0     False     False  False  False   False  
1     False      True  False   True   False  
2      True      True  False   True   False  
3     False     False  False  False    True  
4     False     False  False  False    True  
5     False     False  False  False    True  
6     False      True  False  False   False  
7      True     False   True  False   False  
8     False     False   True   True   False  
9     False     False  False  False    True  
10    False     False  False  False   False  
11    False     False   True   True   False  
12    False     False   True  False   False  
13    False      True  False  False   False  
14     True      True  False  False   False  
15    False     False  False  False    True  
16    False     False   True  False    True  
17     True     False   True  False   False  
18    False      True   True  False    True  
19    False     False   True  False    True

In [6]:
# 최소 지지도
min_support = 0.5

# 지지도 계산
ds_freq = apriori(ds_pre,min_support = 0.5, use_colnames = True)
ds_freq.round(3)

support           itemsets
0     0.55          (BOURBON)
1     0.50             (COKE)
2     0.60         (HEINEKEN)
3     0.65           (OLIVES)
4     0.50  (OLIVES, BOURBON)

In [7]:
ds_asso_rule = association_rules(ds_freq,metric="confidence",min_threshold=0.5)
ds_asso_rule.round(3)

antecedents consequents  antecedent support  consequent support  support  \
0    (OLIVES)   (BOURBON)                0.65                0.55      0.5   
1   (BOURBON)    (OLIVES)                0.55                0.65      0.5   

   confidence   lift  leverage  conviction  
0       0.769  1.399     0.142        1.95  
1       0.909  1.399     0.142        3.85

### [(BOURBON),(OLIVES)] Confidence = 0.909 : BOURBON을 구매한 고객은 OLIVES를 구매할 확률이 0.909정도가 된다. 즉, BOURBON을 마시는 날에는 OLIVES를 안주로 거의 살 생각을 한다는 의미다.
### [(OLIVES),(BOURBON)] Cconfidence = 0.769: OLIVES를 구매한 고객은  BOURBON을 구매할 확률이 0.769정도 가 된다. 즉, OLIVES를 샀는데 갑자기 BOURBON을 마시고 싶다는 생각이 드는 것이다.